samples/matter/lock/src/software_images_swapper.cpp

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include "software_images_swapper.h"
+
+#include <pm_config.h>
+
+#include <dfu/dfu_multi_image.h>
+#include <dfu/dfu_target.h>
+#include <dfu/dfu_target_mcuboot.h>
+#include <zephyr/sys/reboot.h>
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_DECLARE(app, CONFIG_MATTER_LOG_LEVEL);
+
+int SoftwareImagesSwapper::Swap(ApplicationImageId applicationId, SoftwareImagesSwapDoneCallback swapDoneCallback)
+{
+	uint32_t app_address;
+	uint32_t app_size;
+	uint32_t net_address;
+	uint32_t net_size;
+	int result;
+
+	if (mSwapInProgress) {
+		return -EALREADY;
+	}
+
+	if (!swapDoneCallback) {
+		return -EINVAL;
+	}
+
+	if (applicationId == ApplicationImageId::Image_1) {
+		app_address = PM_APP_1_CORE_APP_ADDRESS;
+		app_size = PM_APP_1_CORE_APP_SIZE;
+		net_address = PM_APP_1_CORE_NET_ADDRESS;
+		net_size = PM_APP_1_CORE_NET_SIZE;
+	} else if (applicationId == ApplicationImageId::Image_2) {
+		app_address = PM_APP_2_CORE_APP_ADDRESS;
+		app_size = PM_APP_2_CORE_APP_SIZE;
+		net_address = PM_APP_2_CORE_NET_ADDRESS;
+		net_size = PM_APP_2_CORE_NET_SIZE;
+	} else {
+		LOG_ERR("Requested to swap application image with invalid identifier: %d",
+			static_cast<int>(applicationId));
+		return -EINVAL;
+	}
+
+	result = dfu_target_mcuboot_set_buf(mDFUImageBuffer, sizeof(mDFUImageBuffer));
+	if (result != 0) {
+		return result;
+	}
+
+	mSwapInProgress = true;
+
+	result = SwapImage(app_address, app_size, 0);
+	if (result != 0) {
+		goto exit;
+	}
+
+	result = SwapImage(net_address, net_size, 1);
+	if (result != 0) {
+		goto exit;
+	}
+
+	result = dfu_target_schedule_update(-1);
+	if (result != 0) {
+		goto exit;
+	}
+
+	swapDoneCallback();
+
+exit:
+
+	mSwapInProgress = false;
```

Comment:
This should be only cleared on error, otherwise it won't protect from processing another swap event in the middle of factory reset.